### PR TITLE
Argument to turn on pedantic mode 

### LINKED ISSUE #358: Making it easier to use pedantic mode
Should we add an argument to `$check_syntax()` (or even to `$compile()`) for turning on pedantic mode? I know we already can specify it in `stanc_options` but I think most users will stay away from the `stanc_options` argument. But we could add an argument like this 

```r
mod$check_syntax(pedantic=TRUE) # should probably be FALSE by default
```

which would make it much more likely for users to try it out and hopefully that would both benefit users and help get @rybern and others more feedback on the feature. 

Thoughts? 

This is also related to #298 where @mitzimorris rightfully pointed out that the `include_paths` argument is redundant given that it is part of `stanc_options`. I think users will be more likely to take advantage of certain features if we give them ON/OFF switches rather than making them specify compiler flags. But I also understand @mitzimorris's concern about redundancy and not offering too many ways to do the same thing if not necessary.

### PR DESCRIPTION
**Pull request by jgabry**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and agree to license (see below)

#### Summary

Closes #358 

Adds `pedantic` argument to `$compile()` and `$check_syntax()`. 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting 
(this will be you or your assignee, such as a university or company): 
**Columbia University**

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: R/model.R
Comment (jgabry) [inline R/model.R:526]: I rearranged these arguments a bit. I think for check_syntax `quiet` is probably only used by packages like brms so it doesn't need to be the first argument.

File: tests/testthat/test-model-compile.R
Comment (jgabry) [inline tests/testthat/test-model-compile.R:236]: This isn't new I just moved it up in the file to keep the compilation tests separate from check_syntax tests

File: tests/testthat/test-model-compile.R
Comment (jgabry) [inline tests/testthat/test-model-compile.R:275]: This isn't new I just moved it up in the file to keep the compilation tests separate from check_syntax tests

File: tests/testthat/test-model-compile.R
Comment (jgabry) [inline tests/testthat/test-model-compile.R:284]: I removed the pedantic tests here since we have a separate test for pedantic mode below now

**Comment by jgabry** [comment]
For now I just added a `pedantic` argument to `check_syntax()` but I actually think it could be worth adding it to `compile()` also. What do you think?

File: R/model.R
Comment (rok-cesnovar) [reply R/model.R:526]: Good call.

**Comment by rok-cesnovar** [comment]
> For now I just added a pedantic argument to check_syntax() but I actually think it could be worth adding it to compile() also. What do you think?

I agree with adding it to compile()

**Review by rok-cesnovar — APPROVED** [review]
Looks good. I would add this to compile as well.

**Comment by jgabry** [comment]
Ok thanks, I'll add it to compile today if I have time. If not then one of these next few days.

**Comment by jgabry** [comment]
Ok I just added the argument also for the compile method, however it doesn't work if the executable is already up to date: 

```r
mod <- cmdstan_model(..., pedantic=TRUE) # will show pedantic warnings
mod <- cmdstan_model(..., pedantic=TRUE) # calling it again will just say up to date and not show warnings
```

@rok-cesnovar Do you think that's OK or should we maybe force recompile if pedantic is TRUE?

**Comment by rok-cesnovar** [comment]
> Do you think that's OK or should we maybe force recompile if pedantic is TRUE?

I think that is fine. I don't like forcing recompile, The only other solution would be to call `check_syntax `with pedantic in that case. check syntax is instant.

**Comment by jgabry** [comment]
> The only other solution would be to call `check_syntax `with pedantic in that case. check syntax is instant.

Yeah that's a possibility. We could do it automatically but maybe we can just recommend to the user to call check_syntax() if they want to use pedantic mode for a model that's already compiled? I could add something in the doc about that.

**Comment by jgabry** [comment]
I just pushed a commit mentioning that in the doc, but let me know if you prefer something different and I can change it. Otherwise I think this PR is ready.

**Comment by jgabry** [comment]
Actually this is going to conflict with @mike-lawrence's PR #306 so let's hold off on merging this until we sort things out.

**Comment by rok-cesnovar** [comment]
> I just pushed a commit mentioning that in the doc, but let me know if you prefer something different and I can change it.

Cool with that.

> Actually this is going to conflict with @mike-lawrence's PR #306 so let's hold off on merging this until we sort things out.

That PR has other conflicts we need to sort out, so we can go ahead and merge this once tests pass.

**Comment by mike-lawrence** [comment]
> That PR has other conflicts we need to sort out, so we can go ahead and merge this once tests pass.

I concur!

**Comment by jgabry** [comment]
Ok thanks!

**Comment by rok-cesnovar** [comment]
I now realized we could alao use this in the vignette: https://mc-stan.org/cmdstanr/articles/cmdstanr-internals.html

No rush though.

**Comment by jgabry** [comment]
Good point. That's a quick change so I can do that now.

**Pull request merged**
